### PR TITLE
Corect filtering of OpenStack LBaaS resources to delete

### PR DESCRIPTION
**Pull request by kiall**
Neutron's API ignores unknown parameters. When listing pools etc, K8
attempts to filter on "LoadBalancerID", which is not a valid filter.
As such, it is ignored by Neutron, and a list of all pools is
returned. K8 then proceeds to delete each of the pools.

Instead, we now double check the resources really belong to the LB
we're trying to delete.

Fixes issue #33759

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33768)

### CODE REVIEW

**Comment by dagnello** [comment]
LGTM

**Comment by dagnello** [comment]
@anguslees can we please get tests run on this?

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 1]
```diff
@@ -778,7 +800,13 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
 		}
 		for _, p := range poolsList {
 			for _, l := range p.Listeners {
-				lbPools[l.ID] = p
+				// Double check this Pool belongs to the LB we're deleting. Neutron's API filtering
+				// can't be counted on in older releases (i.e Liberty).
+				for _, val := range lbListeners {
+					if val.ID == l.ID {
+						lbPools[l.ID] = p
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: `break`

**Review by anguslees — COMMENTED** [review]
ok to test

**Comment by Crapworks** [comment]
What is the state here? We have several Kubenetes Cluster on OpenStack, everyone of them has several different customers who are working on the. In the current state, all of our services go down immediately when a single customer deletes one of his services. That causes some severe outages. I'm not sure this issue is getting the attention it deserves.

**Comment by anguslees** [comment]
@k8s-bot ok to test

**Comment by anguslees** [comment]
@kiall: `./hack/verify-gofmt.sh`

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 2]
```diff
@@ -760,8 +776,14 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
...
 		for _, l := range listenersList {
-			key := portKey{Protocol: l.Protocol, Port: l.ProtocolPort}
-			lbListeners[key] = l
+			for _, lb := range l.Loadbalancers {
+				// Double check this Listener belongs to the LB we're updating. Neutron's API filtering
+				// can't be counted on in older releases (i.e Liberty).
+				if loadbalancer.ID == lb.ID {
+					key := portKey{Protocol: l.Protocol, Port: l.ProtocolPort}
+					lbListeners[key] = l
+				}
```
Comment (kiall) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 1 above)
Comment (kiall) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done

**Comment by kiall** [comment]
@k8s-bot verify test this

**Comment by kiall** [comment]
@k8s-bot gci gke e2e test this

Looks like a CI fail to me

**Comment by kiall** [comment]
@k8s-bot gci gke e2e test this

**Comment by anguslees** [comment]
@k8s-bot gci gke e2e test this

**Comment by xsgordon** [comment]
"No cluster named 'e2e-gke-agent-pr-5-0' in k8s-jkns-pr-gci-gke."

Maybe that means something to someone else... :P

**Comment by grahamhayes** [comment]
@k8s-bot gci gke e2e test this

**Comment by kiall** [comment]
Anyone able to LGTM? :)

**Comment by rootfs** [comment]
@mikedanese ?

**Comment by grahamhayes** [comment]
@k8s-bot unit test this

**Comment by kiall** [comment]
@anguslees another LGTM please? finally actually passing the tests this change hasn't touched ;)

**Pull request merged**